vllm/v1/worker/gpu_model_runner.py

```diff
@@ -2165,6 +2157,80 @@ def rand_input_ids() -> torch.Tensor:
             yield
             input_ids.fill_(0)
 
+    def _get_modality_with_max_tokens(self):
+        # NOTE: Currently model is profiled with a single non-text
+        # modality with the max possible input tokens even when
+        # it supports multiple.
+        max_tokens_by_modality_dict = self.mm_registry \
+            .get_max_tokens_per_item_by_nonzero_modality(self.model_config)
+        dummy_data_modality, max_tokens_per_mm_item = max(
+            max_tokens_by_modality_dict.items(), key=lambda item: item[1])
+
+        return dummy_data_modality, max_tokens_per_mm_item
+
+    def _get_mm_dummy_params(self):
+        (
+            dummy_data_modality,
+            max_tokens_per_mm_item,
+        ) = self._get_modality_with_max_tokens()
+
+        # Check how many items of this modality can be supported by
+        # the encoder budget.
+        encoder_budget = min(self.max_num_encoder_input_tokens,
+                             self.encoder_cache_size)
+
+        max_num_mm_items_encoder_budget = encoder_budget // \
+            max_tokens_per_mm_item
+
+        # Check how many items of this modality can be supported by
+        # the decoder budget.
+        max_mm_items_per_req = self.mm_registry.get_mm_limits_per_prompt(
+            self.model_config)[dummy_data_modality]
+
+        # NOTE: We do not consider max_num_batched_tokens on purpose
+        # because the multimodal embeddings can be generated in advance
+        # and chunked prefilled.
+        max_num_mm_items_decoder_budget = self.max_num_reqs * \
+            max_mm_items_per_req
+
+        max_num_mm_items = max(
+            1,
+            min(max_num_mm_items_encoder_budget,
+                max_num_mm_items_decoder_budget))
+
+        logger.info(
+            "Encoder cache will be initialized with a budget of %s tokens, "
+            "and profiled with %s %s items of the maximum feature size.",
+            encoder_budget, max_num_mm_items, dummy_data_modality)
+
+        return dummy_data_modality, max_mm_items_per_req, max_num_mm_items
+
+    def _get_mm_dummy_batch(
+        self,
+        modality: str,
+        processor_batch_size: int,
+        model_batch_size: int,
+    ) -> BatchedTensorInputs:
```

Comment:
good refactor,  I was wondering why this wasn't in the gpu runner already as I was porting it to tpu :)